### PR TITLE
Yieni- Space

### PR DESCRIPTION
**Pull request by yieknee**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | Top-10, which created the top 10 for my homepage by finding each media in the category, sorting them by vote count from greatest to least, then selecting 10. 
Describe how you approached testing that model method. What edge cases did you come up with? | If parameters for user were nil the user couldn't get created.
What are `session` and `flash`? What is the difference between them? | Session tracks when a user when they are signed in using cookies and the session ends when they are signed out. Flash is a message that can pop up when you want it to for success, error, and warnings. 
What was one thing that you gained more clarity on through this assignment? | How to create relationships in rails: many to many, one to many etc.
What is the Heroku URL of your deployed application? | https://yienis-media-ranker.herokuapp.com/

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

### CODE REVIEW

File: app/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+background-color: rgb(25, 112, 103);
+border-color: rgb(25, 112, 103);
+}
+
+ul {
+margin-block-start: 1em;
+margin-block-end: 1em;
+margin-inline-start: 0px;
+margin-inline-end: 0px;
+padding-inline-start: 40px;
```
Comment (kaidamasaki) [inline app/assets/stylesheets/application.scss:76]: These are experimental properties.  You should use regular `margin` and `padding` rules instead of these.

Were these used in the demo app?  (If they were then I need to update the demo app.)

File: app/assets/stylesheets/application.scss [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+}
+
+li {
+display: list-item;
+// text-align: -webkit-match-parent;
+}
+
+.btn-primary {
+background-color: #26A69A;
+color: white;
```
Comment (kaidamasaki) [inline app/assets/stylesheets/application.scss:86]: Generally hex codes are preferred to color names for cross browser consistency (`white` and `black` are consistent though so it's not a big deal, just a recommendation).

File: app/assets/stylesheets/application.scss [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+font-family: "Gudea", sans-serif;
+}
+
+main {
+padding: 2rem;
+}
+
+h2{
+  color: #26A69A;
+}
```
Comment (kaidamasaki) [inline app/assets/stylesheets/application.scss:41]: Inconsistent indentation.

File: app/models/vote.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+class Vote < ApplicationRecord
+  belongs_to :user
+  belongs_to :work
+
+  validates :work_id, uniqueness:{scope: :user_id, message: "User: Has already voted for this work"}, presence: :true
```
Comment (kaidamasaki) [inline app/models/vote.rb:5]: Nice use of a scope!

File: app/models/work.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
+class Work < ApplicationRecord
+  has_many :votes, dependent: :destroy
+  has_many :users, through: :votes
+
+  def self.spotlight
+    works = Work.all
+    max = 0
+    works.each do |work|
+      max = work.votes.count if work.votes.count > max
+    end
```
Comment (kaidamasaki) [inline app/models/work.rb:10]: You can just use `max` here.  It accepts a block.

```suggestion
    max = works.max { |work| work.votes.count }
```

File: app/models/work.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    return works.select{|work| work.votes.count == max }.sample
+  end
+
+  def self.top_ten(category: category)
+    works = Work.where(category: category)
+    top_ten = works.sort_by{|work| -work.votes.count}.take(10)
```
Comment (kaidamasaki) [inline app/models/work.rb:16]: It might be clearer to use `reverse` here instead of doing a sort on the negative count.

```suggestion
    top_ten = works.sort_by{|work| work.votes.count}.reverse.take(10)
```

File: app/views/homepages/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        Top Movies
+      </h2>
+      <% movies = Work.top_ten(category:'movie') %>
+      <% movies.each do |movie| %>
+        <ul class="list-group top-ten__list">
+          <li class="list-group-item">
+            <h4>
+              <%= link_to movie.title, work_path(movie.id)   %>
+              <small class="top-ten__creator">by <%= movie.creator %></small>
+            </h4>
```
Comment (kaidamasaki) [inline app/views/homepages/index.html.erb:27]: You shouldn't skip from a `<h2>` to a `<h4>` without having a `<h3>` in between.  

If you got this from the demo app I'll update it to fix this.  (I made a Trello card to go over the demo solution once I finish all of my grading.)

File: app/views/homepages/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+            <p><%= movie.votes.count%> <%= movie.votes.count > 1? 'votes' : 'vote'%></p>
```
Comment (kaidamasaki) [inline app/views/homepages/index.html.erb:28]: You repeat this code a few times.  This is the kind of thing that would be nice to put in a view helper.

You could potentially just do something like:

```ruby
  def show_votes(count)
    return content_tag(:p, "#{count} #{count > 1? 'votes' : 'vote'}")
  end
```

```suggestion
            <%= show_votes(movie.votes.count) %>
```

File: test/controllers/works_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+  describe "show" do
+    it "responds with success when showing an existing valid work" do
+
+      get(work_path(@work.id))
+      must_respond_with :success
+    end
+
+    it "responds with 404 with an invalid work id" do
+      get(work_path(00000000))
```
Comment (kaidamasaki) [inline test/controllers/works_controller_test.rb:34]: In Ruby `00000000` is the same as `0`.

```suggestion
      get(work_path(0))
```

File: test/models/vote_test.rb [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+  end
+
+  describe 'validations' do
+
+    it 'is valid when user_id and work_id are present' do
+      result = @vote.valid?
+      expect(result).must_equal true
+    end
+
+    it 'is invalid when votename is not present' do
```
Comment (kaidamasaki) [inline test/models/vote_test.rb:18]:
```suggestion
    it 'is invalid when user is not present' do
```

**Comment by kaidamasaki** [comment]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | Can't edit all fields.
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | Not sorted by vote.
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | Only 18 commits.  Please commit more frequently in the future.  This will help with group projects.

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | No filter for finding logged in user.
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | (Didn't use `session`.)

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | No `votes.yml`.
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention |
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention |





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation |
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | 
| Logical/Organized | ✅

**Review by kaidamasaki — COMMENTED** [review]
Great job!  Here are some tips to clean your code up going forward. 😃
